### ISSUE TITLE
Using Dockerfile to build devcontainer with Podman causes VSCode to hang

### ISSUE THREAD

**Issue by awesomekyle** [microsoft/vscode-remote-release#9748]
When creating a devcontainer from a Dockerfile using Podman, VSCode hangs because Podman doesn't pull from `localhost/` when using the `--platform` build arg. Because the terminal Dev Container terminal window can't be interacted with at all, this means VSCode hangs forever and the only way to stop it is to kill VSCode.

The Dockerfile builds correctly ("Successfully tagged localhost/vsc-podman....") and then VSCode does some additional setup using the `/tmp/devcontainercli-kyle/updateUID.Dockerfile-0.58.0` file, which has the following first two lines:

```
ARG BASE_IMAGE
FROM $BASE_IMAGE
...
```

`BASE_IMAGE=vsc-podman-devcontainer-41....53dbb` is being set in the `podman build` command. The issue is that `podman build` prompts for the registry as shown in the image below:

![Screenshot from 2024-04-04 19-26-22](https://github.com/microsoft/vscode-remote-release/assets/303261/38eef4fa-775a-41d5-b652-68d04ccb2172)

**.devcontainer/devcontainer.json**

If I use the `image` option, the container is created correctly. It only fails when `BASE_IMAGE` is a local image.

```
{
  "name": "Alpine",

  // "image": "mcr.microsoft.com/devcontainers/base:alpine-3.19",
  "build": { "dockerfile": "Dockerfile" },

  "workspaceMount": "",
  "runArgs": [
    "--userns=keep-id",
    "--volume=${localWorkspaceFolder}:/workspaces/${localWorkspaceFolderBasename}:z"
  ],
  "containerUser": "vscode"
}

```

**.devcontainer/Dockerfile**
```
FROM mcr.microsoft.com/devcontainers/base:alpine-3.19
```

**Workaround:**

* If `/etc/containers/registries.conf` only has one `unqualified-search-registries`, podman won't prompt for a selection so the container creation will work correctly.
* ~~After it fails, manually run the command that hung, choosing any prompted registry (it doesn't matter, Podman pulls locally regardless).~~ **Edit:** After further testing this isn't true. VSCode tries to rebuild it regardless.
* Push the base container to a registry instead of using the `build` option. **Edit:** Using `features` also causes new containers to be built, so this workaround only works if you don't use devcontainer features.

**Comment by chrmarti** [comment]
Is this a bug in Podman? You mention `--platform`, does it not ask for the registry without `--platform`?

**Comment by awesomekyle** [comment]
I'm not sure if this is a bug, I haven't yet filed a bug with Podman. But that's correct, if `--platform` is not passed in, Podman doesn't prompt for anything. This could be intentional behavior in Podman. If you don't specify a platform, it will automatically use `localhost` because it's the correct platform by definition.

Testing a bit more locally this also means that `features` can't be used either because they also build a new container dynamically. The following `devcontainer.json` will cause a hang when running with Podman on Fedora with `"dev.containers.dockerPath": "podman",`.

```
{
  "name": "Alpine",

  "image": "mcr.microsoft.com/devcontainers/base:alpine-3.19",
  // "build": { "dockerfile": "Dockerfile" },

  "workspaceMount": "",
  "runArgs": [
    "--userns=keep-id",
    "--volume=${localWorkspaceFolder}:/workspaces/${localWorkspaceFolderBasename}:z"
  ],
  "containerUser": "vscode",
  "features": {
    "ghcr.io/wxw-matt/devcontainer-features/command_runner:0": {}
  }
}
```

After testing more I updated my original post; manually running the failed Podman command doesn't work, VSCode continues trying to build the container.

**Comment by awesomekyle** [comment]
In case any body else happens to run into this, you can use the `CONTAINERS_REGISTRIES_CONF` environment variable to override the registries file:

**~/.local/bin/podman-devcontainers:**

```
#!/bin/bash
CONTAINERS_REGISTRIES_CONF="$HOME/.config/devcontainers-registries.conf" podman "$@"
```

**~/.config/devcontainers-registries.conf:**

```
unqualified-search-registries = ["docker.io"]
```

And then use `"dev.containers.dockerPath": "/home/username/.local/bin/podman-devcontainers",` in VSCode's settings. This will then ensure Podman only has one registry to pull from and doesn't hang.

@chrmarti, at the very least there should be a way to cancel the container creation process. If the process fails due to an error, VSCode prompts the user what to do. In this case there's no option other than killing VSCode entirely.

**Comment by lgonzalezsa** [comment]
I was having same issue described here and reversed to v0.348.0 of the extension and now works fine with podman
Issue with v0.354.0:
![image](https://github.com/microsoft/vscode-remote-release/assets/704686/f795daff-6f58-4fae-b6d8-214ee4cf26d3)

**Comment by chrmarti** [comment]
I think https://github.com/devcontainers/cli/pull/746 might have changed it, so we now always pass `--platform`, also when the platform matches the local hardware.

**Comment by pjlammertyn** [comment]
@awesomekyle correct restricting the number of registries  is a workaround

```bash
mkdir $HOME/.config/containers
cat <<EOF > $HOME/.config/containers/registries.conf
unqualified-search-registries = ['docker.io']
EOF
```

**Comment by soiamsoNG** [comment]
@chrmarti 
encounter same issue and after some search get below

podman tag behavior is not as docker
podman will alwasy add **localhost/** before tag  of image without specify registry part
even you are not hosting a local registry server.

but BASE_IMAGE miss **localhost/** before it.

**Comment by chrmarti** [comment]
Not sure what the correct fix is here. It's unfortunate that a change on our side broke things, but I also think adding `--platform` shouldn't suddenly require user input. I'm leaning towards considering this a Podman issue.

**Comment by pgourinchas** [comment]
Is it possible to prefix the BASE_IMAGE argument with localhost/ when the docker path setting contains podman ?
![image](https://github.com/microsoft/vscode-remote-release/assets/6492604/8c0c96c4-1827-4714-8ccb-8a6ea4c94c77)

**Comment by chrmarti** [comment]
This fix is available with Dev Containers 0.360.0-pre-release. Let us know if this works, thanks!

**Comment by lgonzalezsa** [comment]
I was testing switching to Dev Containers 0.360.0-pre-release.

I am getting an error seems related now to trying to reach a hub that does not exist `hub.docker.corp.net` 
Let me know which logs you might need to review the problem or if I should post on another issue.

```bash
[2024-04-19T13:58:22.627Z] STEP 1/9: FROM localhost/hub.docker.corp.net/my_repo/base:latest
[2024-04-19T13:58:22.631Z] Trying to pull localhost/hub.docker.corp.net/my_repo/base:latest...
[2024-04-19T13:58:22.634Z] WARN[0000] Failed, retrying in 2s ... (1/3). Error: initializing source docker://localhost/hub.docker.corp.net/my_repo/base:latest: pinging container registry localhost: Get "https://localhost/v2/": dial tcp [::1]:443: connect: connection refused 
[2024-04-19T13:58:24.643Z] WARN[0002] Failed, retrying in 2s ... (2/3). Error: initializing source docker://localhost/hub.docker.corp.net/my_repo/base:latest: pinging container registry localhost: Get "https://localhost/v2/": dial tcp [::1]:443: connect: connection refused 
[2024-04-19T13:58:26.646Z] WARN[0004] Failed, retrying in 2s ... (3/3). Error: initializing source docker://localhost/hub.docker.corp.net/my_repo/base:latest: pinging container registry localhost: Get "https://localhost/v2/": dial tcp [::1]:443: connect: connection refused 
[2024-04-19T13:58:28.649Z] Error: creating build container: initializing source docker://localhost/hub.docker.corp.net/my_repo/base:latest: pinging container registry localhost: Get "https://localhost/v2/": dial tcp [::1]:443: connect: connection refused
```

**Comment by chrmarti** [comment]
Thanks @lgonzalezsa!

I need to add a check if there is already a registry name since we go through the same code path with that. PR is out: https://github.com/devcontainers/cli/pull/810

**Comment by damaestro** [comment]
Using the pre-release channel resolved this issue for me.

**Comment by chrmarti** [comment]
@lgonzalezsa The updated fix is available with Dev Containers 0.361.0-pre-release. Please give that a try, thanks!

**Comment by lgonzalezsa** [comment]
Awesome! works for me!
